composer/datasets/hf_streaming.py

```diff
@@ -0,0 +1,243 @@
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install mosaicml-composer[nlp]`')
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(self.tokenizer_name)  #type: ignore (thirdparty)
+        self.config = transformers.AutoConfig.from_pretrained(self.tokenizer_name)  #type: ignore (thirdparty)
+
+        # Load and shard dataset
+        text_dataset = datasets.load_dataset(path=self.name, name=self.config_name, split=self.split, streaming=True)
+        text_dataset = self._shard_dataset(text_dataset)
+        assert isinstance(text_dataset, datasets.IterableDataset)
+
+        # Shuffle
+        if self.shuffle:
+            text_dataset = text_dataset.shuffle(buffer_size=self.shuffle_buffer_size, seed=self.seed)
+
+        # Map text samples to token samples
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over tokenized samples
+        text_sample_batch_size = 1000
+        token_dataset = text_dataset.map(
+            self._tokenize,
+            batched=True,
+            batch_size=text_sample_batch_size,
+        )
+
+        if self.group_method != "truncate":
+            # Map variable-length token samples to fixed-length token samples
+            # NOTE: Mapping is executed in batched mode for better CPU utilization,
+            # but the returned dataset is still an iterable over tokenized samples.
+            # NOTE: Depending on the 'group_method', this step may alter the number of
+            # token samples in the dataset, and may mix neighboring token samples together.
+            token_sample_batch_size = 1000
+            token_dataset = token_dataset.map(
+                self._group_tokens,
+                batched=True,
+                batch_size=token_sample_batch_size,
+            )
+
+        # Limit the number of post-processed samples
+        num_samples_per_device = self._get_approx_num_samples_per_device()
+        token_dataset = token_dataset.take(num_samples_per_device)
+
+        # HACK: create a SizedIterableDataset
+        sized_iterable_dataset = SizedIterableDataset(token_dataset, num_samples_per_device)
+
+        # Get collate_fn
+        if self.tokenizer_name in ["gpt2"]:
+            # Annoying but GPT2 tokenizer has no padding token which causes bugs
+            collate_fn = transformers.default_data_collator
+        else:
+            collate_fn = transformers.DataCollatorForLanguageModeling(tokenizer=self.tokenizer,
+                                                                      mlm=self.use_masked_lm,
+                                                                      mlm_probability=self.mlm_probability)
+        # Return DataSpec
+        return DataSpec(dataloader=dataloader_hparams.initialize_object(
+            dataset=sized_iterable_dataset,
```

Comment:
Can you pass num_samples and num_tokens into the DataSpec?

Reply:
I'm not sure if we are quite ready yet... there's still some discussion with NLP team as to what constitutes a sample / a token when there is padding and concatenation involved. I can bring it up with relevant team members and make a follow up PR.